Source/Dafny/Compilers/SinglePassCompiler.cs

```diff
@@ -1714,10 +1715,18 @@ protected class NullClassWriter : IClassWriter {
       // One of the limitations in many target language encodings are restrictions to instance members. If an
       // instance member can't be directly expressed in the target language, we make it a static member with an
       // additional first argument specifying the `this`, giving it a `CustomReceiver`.
-      return !member.IsStatic
-             && (member.EnclosingClass is NewtypeDecl
-                 || (member.EnclosingClass is TraitDecl
-                     && member is ConstantField { Rhs: { } } or Function { Body: { } } or Method { Body: { } }));
+      if (member.IsStatic) {
+        return false;
+      } else if (member.EnclosingClass is NewtypeDecl) {
+        return true;
+      } else if (member.EnclosingClass is TraitDecl) {
+        return member is ConstantField { Rhs: { } } or Function { Body: { } } or Method { Body: { } };
+      } else if (member.EnclosingClass is DatatypeDecl datatypeDecl) {
+        // An undefined value "o" cannot use this o.F(...) form in most languages.
+        return datatypeDecl.Ctors.Any(ctor => ctor.IsGhost);
+      } else {
+        return false;
+      }
```

Comment:
Consider
```C#
      if (member.IsStatic) {
        return false;
      }

      return member.EnclosingClass switch {
        NewtypeDecl => true,
        TraitDecl => member is ConstantField { Rhs: { } } or Function { Body: { } } or Method { Body: { } },
        DatatypeDecl datatypeDecl => datatypeDecl.Ctors.Any(ctor => ctor.IsGhost),
        _ => false
      };
```
or
```C#
      return !member.IsStatic
             &&     (member.EnclosingClass is NewtypeDecl
                 || (member.EnclosingClass is TraitDecl
                     && member is ConstantField { Rhs: { } } or Function { Body: { } } or Method { Body: { } })
                 || (member.EnclosingClass is DatatypeDecl datatypeDecl && datatypeDecl.Ctors.Any(ctor => ctor.IsGhost)));
```

Reply:
Is the second version easier to read? I'm hesitant about expressions that require parenthesis.

Reply:
I'll stay with the original.